### PR TITLE
libsubprocess: do not spin on large lines

### LINKED ISSUE #6262: flux cannot handle very large lines
While working on something else, I noticed

`t/shell/lptest 4000000 1000` worked as a job under flux but `t/shell/lptest 5000000 1000` did not.  The latter would just hang and produce no output.

Only later did I realize I messed up.  I should have done `t/shell/lptest 1000 4000000` (the first number is the length of a line, the second is the number of lines).   So I was inadvertently creating > 4 meg lines.

The > 4 meg line probably crossed some internal buffer / boundary somewhere, leading to problems.  But no error messages here showing up.

### PR DESCRIPTION
**Pull request by chu11**
Problem: In several cases, libsubprocess hangs/spins can occur if the internal output buffer is full.  For example, if subprocess output is line buffered and a single line exceeds the buffer size, the buffer can never be emptied because output callbacks are never called (i.e. the buffer never contains a line).

Other situations can exist if the user simply does not read data when it becomes available.

Solution: Handle full output buffers with two special cases

- if output is line buffered and the buffer is full AND no line exists, call the output callback for the user to get the current data.
  flux_subprocess_read_line() and similar functions will return data that is not a full line.

- if the buffer is at capacity and the user elected to not read anything in the output callback, drop the data.  The internal assumption is that a user must read data that is given to them at that point in time.

Fixes #6262

### CODE REVIEW

**Comment by garlick** [comment]
Nice!  does this fix #4572 also?

**Comment by chu11** [comment]
Hmm, I don't think so.  The `Value too large for defined data type` suggests it might be an issue more similar to https://github.com/flux-framework/flux-core/issues/6256

File: src/common/libsubprocess/ev_fbuf_read.c [hunk 1]
```diff
@@ -105,6 +110,31 @@ static void check_cb (struct ev_loop *loop, ev_check *w, int revents)
...
+             * "fallout" from a watcher just (effectively) exiting out
+             * of the blue.  From caller perspectives, it may have
+             * exited cleanly.
+             *
+             * we choose to dump the buffer contents instead.
+             * Unfortunately this leads to loss of data and no error
+             * message.  However in authors opinion, it is a "cleaner"
+             * fallout.
+             *
+             */
```
Comment (garlick) [inline src/common/libsubprocess/ev_fbuf_read.c:135]: Pondering this a bit, it seems like this ev watcher ought to behave similar to the libev ev_io watcher:

http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#code_ev_io_code_is_this_file_descrip

That is, it should be level triggered (keep calling callback while data is available) and should be stoppable in the watcher callback.  This change would drop data if the watcher chose to stop itself rather than consume the data so I think it would not be correct.

It may be good to add some inline documentation though (I guess in `fbuf_watcher.h`?) that the watchers are level triggered and that the read watcher must either read some data or stop the watcher to avoid unnecessary trips through the event loop without progress.

Also, maybe add documentation in subprocess.h that the output callbacks should either consume data or call `flux_subprocess_stream_stop()`.

**Review by garlick — COMMENTED** [review]
See inline comment - it seems like the second problem may be more of a documentation one?

File: src/common/libsubprocess/ev_fbuf_read.c (hunk 1 above)
Comment (garlick) [reply src/common/libsubprocess/ev_fbuf_read.c:135]: Maybe this PR should just ensure that, when line buffered and the buffer is full, the callback gets even there isn't a line terminator, and level triggering documentation (and any fixes for in-tree inappropriate uses, if any) could go in a separate PR?

Comment (chu11) [reply src/common/libsubprocess/ev_fbuf_read.c:135]: > Maybe this PR should just ensure that, when line buffered and the buffer is full, the callback gets called even there isn't a line terminator, and level triggering documentation (and any fixes for in-tree inappropriate uses, if any) could go in a separate PR?

Let me try again to split it up.  I think the reason I didn't for this PR is b/c a user not being able to read a buffer without a line  == a user not reading data b/c they "forgot to".  So that's why I ended up putting it all together.  It led to some testing difficulties otherwise.  But I'll give it another shot.

**Comment by garlick** [comment]
I was questioning the correctness of the proposed change for "forgot to".

File: src/common/libsubprocess/remote.c [hunk 2]
```diff
@@ -438,6 +442,13 @@ static int remote_output_buffered (flux_subprocess_t *p,
     if (data && len) {
         int tmp;
 
+        /* In the event the buffer is full, the `fbuf_write()` will
+         * fail.  Call user callback to give them a chance to empty
+         * buffer.  If they don't, we'll hit error below.
+         */
+        if (!fbuf_space (c->read_buffer))
+            c->output_cb (c->p, c->name);
+
```
Comment (garlick) [inline src/common/libsubprocess/remote.c:451]: Was going to comment that throwing an error here is not consistent with allowing the user to stop the stream from the callback, but I see the stream_start/stop functions are noted to be for local processes only.

We can fix that once we have credit based flow control since the remote will never send more data than we have room to put in the buffer.

**Comment by chu11** [comment]
re-pushed, this PR is now limited to just the line buffering corner case

File: src/common/libsubprocess/test/stdio.c [hunk 3]
```diff
@@ -1395,6 +1473,8 @@ int main (int argc, char *argv[])
     test_stream_start_stop_initial_stop (r);
     diag ("stream_start_stop_mid_stop");
     test_stream_start_stop_mid_stop (r);
+    diag ("overflow_output_buffer");
+    test_overflow_output_buffer (r);
```
Comment (garlick) [inline src/common/libsubprocess/test/stdio.c:1477]: Suggestoin: rename this test to something like `test_long_lines()` or similar since the point is to demonstrate that the buffer does _not_ overflow when it gets a long line.

**Review by garlick — COMMENTED** [review]
Looking good!  I have a few comments that are mostly trivial.

I did want to manually kick it a bit with `lptest` and see how things go so I'll save approval for after that test.

**Review by garlick — APPROVED** [review]
I did some testing with `flux exec` running lptest with various line sizes and lengths, and comparing the output to the same command run locally.  I even tried 4G worth of (4MB-500B) lines!  No issues.

I then repeated some of these tests with the data going to a job's stdout.  Also no issues.

Approving!  Nice work.

File: src/common/libsubprocess/remote.c (hunk 2 above)
Comment (chu11) [reply src/common/libsubprocess/remote.c:451]: the callback placement here is unfortunate, the reason was the libev did not call things in the order I expected.   I expected:

```
output - put data in buffer, start output prep/check
check_cb - call output callback since there's data in the buffer
<start next iteration of libev loop>
```

but what happened was

```
output - put data in buffer, start output prep/check
<start next iteration of libev loop>
prep_check - see data in output buffer, start idle
output - want to put more data in buffer, hit EOVERFLOW
check_cb - this is never reached because of error above
```

the fact I just started the prep/check means check isn't called in the current iteration.

Comment (chu11) [reply src/common/libsubprocess/remote.c:451]: > We can fix that once we have credit based flow control since the remote will never send more data than we have room to put in the buffer.

The work on #6291 is only for stdin since that's the specific case brought up by the user.  But yeah, for output we should add that as a todo as well.

Comment (garlick) [reply src/common/libsubprocess/remote.c:451]: On the order of events, should the output watcher be stopped when the buffer is full then, and restarted when it's not?

Comment (garlick) [reply src/common/libsubprocess/remote.c:451]: I think that's sort of like putting an infinite max size on the cbuf - it allows the sender to cause the receiver to use up an unchecked amount of memory.  Not great.

Another route might be to expose watcher priorities in our APIs, and then use them to ensure this check watcher runs before anything else.

But maybe the right solution is flow control.

Comment (chu11) [reply src/common/libsubprocess/remote.c:451]: > Another route might be to expose watcher priorities in our APIs, and then use them to ensure this check watcher runs before anything else.

I wasn't aware of the libev priority stuff.  Hmmmm.  I suppose that could be an option, but at this point in time I'm not sure we have a way to add a priority to whatever underneath the covers is calling the flux future's then callback?  So perhaps this is something to simply kick the can down the road.

**Comment by chu11** [comment]
re-pushed with tweaks per comments above

File: src/common/libsubprocess/remote.c (hunk 2 above)
Comment (garlick) [reply src/common/libsubprocess/remote.c:451]: Well we could elevate the priority of just this check watcher to get it to be called before the  check watcher in the future implementation.   Did you want to pause and try that?  I could give you a commit that adds a `flux_watcher_set_priority()` function to cherry pick.  Just as an experiment?

Comment (garlick) [reply src/common/libsubprocess/remote.c:451]: Untested but here it is: 1932ce57c5c167385759e4fef7cd22294f5d25ca

Just call `flux_watcher_set_priority (check_watcher, 1);`

That should raise the priority from the default of 0 to 1.  if that doesn't work try 2 :-)  2 is the max.

Comment (chu11) [reply src/common/libsubprocess/remote.c:451]: > Did you want to pause and try that?

As this would involve more than a few line tweak, I'm inclined to merge this PR and experiment with it in a different PR.  But lets log so we don't forget this conversation.

Comment (garlick) [reply src/common/libsubprocess/remote.c:451]: Sure.
